manim/config.py

```diff
@@ -1,326 +1,413 @@
-    dirs.VIDEO_DIR = dir_config["video_dir"]
-    dirs.TEX_DIR = dir_config["tex_dir"]
-    dirs.TEXT_DIR = dir_config["text_dir"]
-
-def register_tex_template(tpl):
-    """Register the given LaTeX template for later use.
-
-    Parameters
-    ----------
-    tpl : :class:`~.TexTemplate`
-        The LaTeX template to register.
-    """
-    constants.TEX_TEMPLATE = tpl
-
-def initialize_tex(config):
-    """Safely create a LaTeX template object from a file.
-    If file is not readable, the default template file is used.
-
-    Parameters
-    ----------
-    filename : :class:`str`
-        The name of the file with the LaTeX template.
-    """
-    filename=""
-    if config["tex_template"]:
-        filename = os.path.expanduser(config["tex_template"])
-    if filename and not os.access(filename, os.R_OK):
-        # custom template not available, fallback to default
-        logger.warning(
-            f"Custom TeX template {filename} not found or not readable. "
-            "Falling back to the default template."
-        )
-        filename = ""
-    if filename:
-        # still having a filename -> use the file
-        constants.TEX_TEMPLATE = TexTemplateFromFile(filename=filename)
-    else:
-        # use the default template
-        constants.TEX_TEMPLATE = TexTemplate()
+
+def _from_command_line():
+    """Determine if manim was called from the command line."""
+    # Manim can be called from the command line in three different
+    # ways.  The first two involve using the manim or manimcm commands
+    prog = os.path.split(sys.argv[0])[-1]
+    from_cli_command = prog in ['manim', 'manimcm']
+
+    # The third way involves using `python -m manim ...`.  In this
+    # case, the CLI arguments passed to manim do not include 'manim',
+    # 'manimcm', or even 'python'.  However, the -m flag will always
+    # be the first argument.
+    from_python_m = sys.argv[0] == '-m'
+
+    return from_cli_command or from_python_m
+
+
+# Config files to be parsed, in ascending priority
+library_wide = os.path.join(os.path.dirname(__file__), 'default.cfg')
+config_files = [
+    library_wide,
+    os.path.expanduser('~/.manim.cfg'),
+    os.path.join(os.getcwd(), 'manim.cfg'),
+]
+
+if _from_command_line():
+    args = _parse_cli(sys.argv[1:])
+    if args.config_file is not None:
+        config_files.append(args.config_file)
+
+else:
+    # In this case, we still need an empty args object.
+    args = _parse_cli([], input=False)
+    # Need to populate the options left out
+    args.file, args.scene_names, args.output_file = '', '', ''
+
+config_parser = configparser.ConfigParser()
+successfully_read_files = config_parser.read(config_files)
+logger.info(f'Read configuration files: {successfully_read_files}')
+
+# this is for internal use when writing output files
+file_writer_config = _parse_file_writer_config(config_parser, args)
+
+# this is for the user
+config = _parse_config(config_parser, args)
+camera_config = config
```

Comment:
I feel like this is asking for trouble. We're exporting two references to the same data structure in different variables, which means changes made with one will silently affect the other. And there's no indication of this given to the user.

But then again, what's the benefit of exporting config and camera_config separately anyway?

Reply:
Yes I was sloppy here. The main reason for me doing this is that in original 3b1b manim there is a `camera_config` sub-dictionary inside the main config dictionary that is sometimes used in a standalone way. So in the past it was a sub-dictionary inside a dictionary that doesn't exist anymore, so we can't add a `DeprecationWarning` anywhere (or can we?). So my idea was to export a `camera_config` dict for semi-backwards compatibility purposes. (And also because `camera_config` is more expressive than just `config`.)

On the issue of exporting two different copies, and on the issue of actually making changes to the global `config` dict: I'm leaning more and more toward exporting a `frozendict` instead of a `dict` since it doesn't make much semantic sense to try to change the global `config` dict in any case. Since `config` will always be module-local, changes made to it will also remain module-local and very very hard to debug in case they happen. If the user wants to refer to the global `config` dict but possibly change its values, that's what the class-local `CONFIG` dicts are for (and will soon be replaced with `attr`s).

So I vote we convert the dicts exported by `config.py` into `frozendict`s and maybe get rid of `camera_config`, though this will make our v1.0.0 not exactly backwards compatible.

Reply:
we can prob keep `camera_config` so long as we don't use it anywhere within the `manim` codebase

Reply:
regarding `frozendict`: there might be usecases where you want to make global changes to things, but we still gotta check that out... we should prob try keeping it as a `dict` for now and discuss a possible change to `frozendict` later 

Reply:
I think the better question is "if the user wants to make a global change to the config variables, how would they do it?". But first, I would like to know which use cases require such a case and determine whether we want to support them.

Reply:
I think using a frozendict is a good workaround, although we should probably remove one of them eventually.

I don't see why someone would want to change the config programmatically rather than just update the data in the .cfg file.

Reply:
they could want to change config mid-animation, i.e. they could want options to vary throughout the video. If, later on, we add color schemes for example, there could be different parts of the Scene with different color schemes, ya get me

Reply:
Even so, it'd make more sense to use a different config file for the part of the video using the different colors rather than typing them into the script directly.

Reply:
If the user wants a major change in the config in between two animations, I think the best way to do that would be to render two different scenes, each using a different config. And then stitching them with a video editor. If they just want to tweak the one option, I think it depends on which option it is. For example, a lot of the `config` keys should I think remain constant e.g. the frame geometry ones `PIXEL_WIDTH`, as well as the directories/paths/filenames. There are only a few config options that I can see being changed mid-scene and those could be handled in other ways. 

(Don't even want to think about changing config mid-animation....)

Reply:
I'm making an executive decision here in the interest of merging this ASAP. I'm going with @PgBiel's suggestion of leaving the exported variables as dicts and opening a new issue to discuss a possible move to frozendict.